pkg/sql/row/writer.go

```diff
@@ -136,31 +136,43 @@ func prepareInsertOrUpdateBatch(
 			if !ok {
 				continue
 			}
+
+			var marshaled roachpb.Value
+			var err error
+			typ := fetchedCols[idx].GetType()
+
 			// Skip any values with a default ID not stored in the primary index,
 			// which can happen if we are adding new columns.
-			if skip := helper.SkipColumnNotInPrimaryIndexValue(family.DefaultColumnID, values[idx]); skip {
-				continue
-			}
-			typ := fetchedCols[idx].GetType()
-			marshaled, err := valueside.MarshalLegacy(typ, values[idx])
-			if err != nil {
-				return nil, err
+			skip, couldBeComposite := helper.SkipColumnNotInPrimaryIndexValue(family.DefaultColumnID, values[idx])
+			if skip {
+				// If the value could be composite, there could be a previous KV, so we
+				// still need to issue a delete.
+				if !couldBeComposite {
+					continue
+				}
+			} else {
+				marshaled, err = valueside.MarshalLegacy(typ, values[idx])
+				if err != nil {
+					return nil, err
+				}
 			}
 
 			// TODO(ssd): Here and below investigate reducing the
 			// number of allocations required to marshal the old
 			// value.
 			var oldVal []byte
 			if oth.IsSet() && len(oldValues) > 0 {
-				old, err := valueside.MarshalLegacy(typ, oldValues[idx])
-				if err != nil {
-					return nil, err
+				if !couldBeComposite || oldValues[idx].(tree.CompositeDatum).IsComposite() {
```

Comment:
I think we can't go wrong with a comment here, because to be honest I'm finding it hard to follow. Let me try to think this through in realtime here in this comment:

We are encoding a single-column, non-family-0 column family.  So, our _expected bytes_ should be:

1. Nil if the value is NULL.
2. Nil is the value is non-NULL but exists _complete_ in the primary key (i.e. is in the primary key and is not a composite value).
3. Non nil otherwise.

Unless I'm reading wrongly, I think this code probably is not quite right for case 1. since I think `TagAndDataBytes` will explode if old was encoded with just nil, so we might have to wrap that in an 'IsPresent()' call.

Reply:
Added a little comment, and wrapped the call to `TagAndDataBytes` in a check for `IsPresent`.

(I'll leave the assertions up to you in your PR.)